File: 0129 - Sum Root to Leaf Numbers/Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        return solve(root, 0);
    }
    
    int solve(TreeNode* head, int sum = 0) {
        if (head == NULL) {
            return {};
        }
        sum = sum * 10 + head->val;
        if (!head->left && !head->right) {
            return sum;
        }
        return solve(head->left, sum) + solve(head->right, sum);
    }
};

File: 0129 - Sum Root to Leaf Numbers/Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
def solve(root, result):
    if root == None:
        return 0
    result = result * 10 + root.val
    if not root.left and not root.right:
        return result
        
    return solve(root.left, result) + solve(root.right, result)


class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        return solve(root, 0)

